server/src/main/java/org/elasticsearch/index/mapper/RootObjectMapper.java

```diff
@@ -400,18 +390,17 @@ private static void validateDynamicTemplate(Mapper.TypeParser.ParserContext pars
             return;
         }
 
-        final XContentFieldType[] types;
-        if (template.getXContentFieldType() != null) {
-            types = new XContentFieldType[]{template.getXContentFieldType()};
-        } else if (template.isRuntimeMapping()) {
-            types = Arrays.stream(XContentFieldType.values()).filter(XContentFieldType::supportsRuntimeField)
-                .toArray(XContentFieldType[]::new);
-        } else {
-            types = XContentFieldType.values();
+        if (parserContext.indexVersionCreated().before(Version.V_8_0_0)) {
+            if (template.pathMatch() == null && template.match() == null && template.getXContentFieldTypes().length == 0) {
+                throw new MapperParsingException("dynamic template [" + template.name() + "] of index " +
+                    parserContext.getIndexSettings().getIndex() + " created before [" + Version.V_8_0_0 + "]" +
+                    " must have [match] or [path_match] or [match_mapping_type] defined");
+            }
         }
```

Comment:
In my opinion we should allow this on old indices the same way we allow adding new field types to old indices. Are we expecting something to break without this validation?

Reply:
I may missing something here but I was under the impression that dynamic template validation happens on every node as we parse the mappings. So, if we accepted the newer format we would end up failing when parsing the mappings on older nodes, and I am not terribly sure what effect that will have. It would be a good idea to test this, even manually, just to validate our assumptions.

Reply:
Also it seems consistent to accept a format when we know it is valid, rather than it causing errors down the line on some shards, again assuming that this is what happens in this case.

Reply:
All nodes hold the cluster state, but we only parse mappings on nodes that hold a shard for the index that has these mappings. And the way that we avoid making everything fail in mixed-version cluster scenarios, is by requiring indices created by version N to be allocated on nodes whose version is >= N. Otherwise we would have the same problem whenever we are adding new field types or new options on existing field types. For new APIs or API parameters like `_pit` or the `dynamic_templates` parameter of bulk requests, we expect users of the cluster to wait until the cluster is fully upgraded before starting using them. This is why our upgrade instructions require upgrading Elasticsearch before Kibana and Beats/Agent.

Reply:
thanks for clarifying, I am aware of how shards are allocated based on their version, which is why it was suggested that we should reject the bulk request preemptively on older indices that don't support the new feature. I like how that approach guarantees consistency, meaning that when we accept the new parameter, we will apply it rather than failing on some shards (those allocated on the older nodes). This sounds like it is a good way to enforce that shippers don't use new features until all nodes are on the same upgraded version. What I was missing is that older indices that are allocated on newer nodes that support the feature will effectively not support the new feature, which I think is too restrictive. I am happy relying on the stream version here, under the assumption that mixed clusters are a temporary situation in which using a new feature is best effort.

Reply:
My comment mentioned the bulk request, while here we were specifically discussing mappings and adding the dynamic template in the newer format. Sorry about the confusion. You asked if we are expecting something to break without this validation: some nodes will fail parsing the mappings because of a format they don't understand, while others will be able to accept it, depending on where shards are allocated. I think what you are hinting at is that, despite this may sound inconsistent, it leads to a consistent behaviour, meaning that if at least one shard is allocated to an older version, the mapping update will be rejected. And that is acceptable in a rolling upgrade scenario, more than having to create a new index to use the new feature. I think I understand better now and I agree.

Reply:
Hi @javanna @jpountz 

Sorry for the late response. I was offline for the last week. I implemented a BWC check using the [min node version](https://github.com/elastic/elasticsearch/pull/69948/commits/aef927e301f68b67ea3f44a3a1cbaee866377b17). However, it is not correct because we [can't](https://github.com/elastic/elasticsearch/blob/a92a647b9f17d1bddf5c707490a19482c273eda3/server/src/main/java/org/elasticsearch/cluster/service/ClusterApplierService.java#L351) retrieve the cluster state from ClusterService while performing a cluster state update. I tried to pass the cluster state around, but the approach is invasive.

I looked at some implementations that introduce new mapping types. We assume that users won't use them until all nodes are upgraded; otherwise, some shards on old nodes will be failed and assigned to new nodes. I pushed https://github.com/elastic/elasticsearch/pull/69948/commits/60d11a4b2bdcd2cc6d1e86ad666449f008b4f140 to relax this validation so users can use this new format with old indices.

Are we okay to get this PR in as other new features? I will spend some time working on this validation later.